File: RandomPointinNonoverlappingRectangles2.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution {
private:
	vector<vector<int> > rects;
public:
    Solution(vector<vector<int> >& rects) {
		this->rects = rects;
    }
    
    vector<int> pick() {
    	int sumArea = 0;
    	vector<int> selected;
    	for(auto rect : rects) {
    		int area = (rect[2] - rect[0] + 1) * (rect[3] - rect[1] + 1);
    		sumArea += area;
    		if(rand() % sumArea < area) selected = rect;
		}
		int x = rand() % (selected[2] - selected[0] + 1) + selected[0];
		int y = rand() % (selected[3] - selected[1] + 1) + selected[1];
		return {x, y};
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<vector<int> > rects;
	for(int i = 0; i < n; i++) {
		vector<int> rect;
		for(int j = 0; j < 4; j++) {
			int num;
			cin>>num;
			rect.push_back(num);
		}
		rects.push_back(rect);
	}
	
	Solution *solution = new Solution(rects);
	
	vector<int> res = solution->pick();
	for(int n : res) cout<<n<<" ";
	cout<<endl;
	
	return 0;
}

File: RandomPointinNonoverlappingRectangles2.py
import random, bisect


class Solution:

    def __init__(self, rects: list):
        self.rects = rects

    def pick(self) -> list:
        sumArea = 0
        select = []
        for x1, y1, x2, y2 in self.rects:
            area = (x2 - x1 + 1) * (y2 - y1 + 1)
            sumArea += area
            if random.randint(1, sumArea) <= area:
                select = [x1, y1, x2, y2]
        return [random.randint(select[0], select[2]), random.randint(select[1], select[3])]


if __name__ == '__main__':
    n = int(input())
    rects = []
    for i in range(n):
        rects.append([int(num) for num in input().split()])

    solution = Solution(rects)
    print(solution.pick())
